Duplication of an element to a mono valued containment feature while updating incoming reference fails

While duplicating a element with incoming derived references and checking the option "Update incoming references" the duplication fails. See stacktrace below. The problem occurs when updating the incoming references. The result of the SetCommand used to attach the element to the model is the container. The current implementation expected to be the child.

To reproduce import the following project:

[ADA - Papaya dependencies (1).zip](https://github.com/user-attachments/files/22661369/ADA.-.Papaya.dependencies.1.zip)

```
java.util.concurrent.ExecutionException: java.lang.ArrayStoreException: org.eclipse.sirius.components.papaya.impl.ContainingLinkImpl
	at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122) ~[na:na]
	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191) ~[na:na]
	at org.eclipse.sirius.components.collaborative.editingcontext.EditingContextEventProcessor.handle(EditingContextEventProcessor.java:229) ~[classes/:na]
	at org.eclipse.sirius.components.collaborative.editingcontext.EditingContextEventProcessorRegistry.lambda$dispatchEvent$1(EditingContextEventProcessorRegistry.java:94) ~[classes/:na]
	at java.base/java.util.Optional.map(Optional.java:260) ~[na:na]
	at org.eclipse.sirius.components.collaborative.editingcontext.EditingContextEventProcessorRegistry.dispatchEvent(EditingContextEventProcessorRegistry.java:94) ~[classes/:na]
	at org.eclipse.sirius.web.application.editingcontext.services.EditingContextDispatcher.dispatchMutation(EditingContextDispatcher.java:94) ~[classes/:na]
	at org.eclipse.sirius.web.application.views.explorer.controllers.MutationDuplicateObjectDataFetcher.get(MutationDuplicateObjectDataFetcher.java:52) ~[classes/:na]
	at org.eclipse.sirius.web.application.views.explorer.controllers.MutationDuplicateObjectDataFetcher.get(MutationDuplicateObjectDataFetcher.java:33) ~[classes/:na]
	at graphql.execution.ExecutionStrategy.invokeDataFetcher(ExecutionStrategy.java:510) ~[graphql-java-24.1.jar:na]
	at graphql.execution.ExecutionStrategy.fetchField(ExecutionStrategy.java:464) ~[graphql-java-24.1.jar:na]
	at graphql.execution.ExecutionStrategy.fetchField(ExecutionStrategy.java:404) ~[graphql-java-24.1.jar:na]
	at graphql.execution.ExecutionStrategy.resolveFieldWithInfo(ExecutionStrategy.java:363) ~[graphql-java-24.1.jar:na]
	at graphql.execution.AsyncSerialExecutionStrategy.resolveSerialField(AsyncSerialExecutionStrategy.java:73) ~[graphql-java-24.1.jar:na]
	at graphql.execution.AsyncSerialExecutionStrategy.lambda$execute$0(AsyncSerialExecutionStrategy.java:57) ~[graphql-java-24.1.jar:na]
	at graphql.execution.Async.eachSequentiallyPolymorphicImpl(Async.java:329) ~[graphql-java-24.1.jar:na]
	at graphql.execution.Async.eachSequentially(Async.java:317) ~[graphql-java-24.1.jar:na]
	at graphql.execution.AsyncSerialExecutionStrategy.execute(AsyncSerialExecutionStrategy.java:52) ~[graphql-java-24.1.jar:na]
	at graphql.execution.Execution.executeOperation(Execution.java:205) ~[graphql-java-24.1.jar:na]
	at graphql.execution.Execution.execute(Execution.java:124) ~[graphql-java-24.1.jar:na]
	at graphql.GraphQL.execute(GraphQL.java:549) ~[graphql-java-24.1.jar:na]
	at graphql.GraphQL.lambda$parseValidateAndExecute$13(GraphQL.java:479) ~[graphql-java-24.1.jar:na]
	at java.base/java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:1187) ~[na:na]
	at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2309) ~[na:na]
	at graphql.EngineRunningState.compose(EngineRunningState.java:87) ~[graphql-java-24.1.jar:na]
	at graphql.GraphQL.parseValidateAndExecute(GraphQL.java:474) ~[graphql-java-24.1.jar:na]
	at graphql.GraphQL.lambda$executeAsync$9(GraphQL.java:434) ~[graphql-java-24.1.jar:na]
	at java.base/java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:1187) ~[na:na]
	at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2309) ~[na:na]
	at graphql.EngineRunningState.compose(EngineRunningState.java:87) ~[graphql-java-24.1.jar:na]
	at graphql.GraphQL.lambda$executeAsync$10(GraphQL.java:423) ~[graphql-java-24.1.jar:na]
	at graphql.EngineRunningState.call(EngineRunningState.java:198) ~[graphql-java-24.1.jar:na]
	at graphql.GraphQL.executeAsync(GraphQL.java:416) ~[graphql-java-24.1.jar:na]
	at graphql.GraphQL.execute(GraphQL.java:359) ~[graphql-java-24.1.jar:na]
	at org.eclipse.sirius.components.graphql.controllers.GraphQLController.performRequest(GraphQLController.java:155) ~[classes/:na]
	at jdk.internal.reflect.GeneratedMethodAccessor47.invoke(Unknown Source) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:569) ~[na:na]
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:258) ~[spring-web-6.2.10.jar:6.2.10]
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:191) ~[spring-web-6.2.10.jar:6.2.10]
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:118) ~[spring-webmvc-6.2.10.jar:6.2.10]
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:991) ~[spring-webmvc-6.2.10.jar:6.2.10]
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:896) ~[spring-webmvc-6.2.10.jar:6.2.10]
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87) ~[spring-webmvc-6.2.10.jar:6.2.10]
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1089) ~[spring-webmvc-6.2.10.jar:6.2.10]
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:979) ~[spring-webmvc-6.2.10.jar:6.2.10]
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1014) ~[spring-webmvc-6.2.10.jar:6.2.10]
	at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:914) ~[spring-webmvc-6.2.10.jar:6.2.10]
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:590) ~[tomcat-embed-core-10.1.44.jar:6.0]
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:885) ~[spring-webmvc-6.2.10.jar:6.2.10]
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:658) ~[tomcat-embed-core-10.1.44.jar:6.0]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:195) ~[tomcat-embed-core-10.1.44.jar:10.1.44]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[tomcat-embed-core-10.1.44.jar:10.1.44]
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:51) ~[tomcat-embed-websocket-10.1.44.jar:10.1.44]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164) ~[tomcat-embed-core-10.1.44.jar:10.1.44]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[tomcat-embed-core-10.1.44.jar:10.1.44]
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100) ~[spring-web-6.2.10.jar:6.2.10]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.2.10.jar:6.2.10]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164) ~[tomcat-embed-core-10.1.44.jar:10.1.44]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[tomcat-embed-core-10.1.44.jar:10.1.44]
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93) ~[spring-web-6.2.10.jar:6.2.10]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.2.10.jar:6.2.10]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164) ~[tomcat-embed-core-10.1.44.jar:10.1.44]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[tomcat-embed-core-10.1.44.jar:10.1.44]
	at org.springframework.web.filter.ServerHttpObservationFilter.doFilterInternal(ServerHttpObservationFilter.java:110) ~[spring-web-6.2.10.jar:6.2.10]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.2.10.jar:6.2.10]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164) ~[tomcat-embed-core-10.1.44.jar:10.1.44]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[tomcat-embed-core-10.1.44.jar:10.1.44]
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201) ~[spring-web-6.2.10.jar:6.2.10]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.2.10.jar:6.2.10]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164) ~[tomcat-embed-core-10.1.44.jar:10.1.44]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[tomcat-embed-core-10.1.44.jar:10.1.44]
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:167) ~[tomcat-embed-core-10.1.44.jar:10.1.44]
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:90) ~[tomcat-embed-core-10.1.44.jar:10.1.44]
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:483) ~[tomcat-embed-core-10.1.44.jar:10.1.44]
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:116) ~[tomcat-embed-core-10.1.44.jar:10.1.44]
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:93) ~[tomcat-embed-core-10.1.44.jar:10.1.44]
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74) ~[tomcat-embed-core-10.1.44.jar:10.1.44]
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:344) ~[tomcat-embed-core-10.1.44.jar:10.1.44]
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:398) ~[tomcat-embed-core-10.1.44.jar:10.1.44]
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:63) ~[tomcat-embed-core-10.1.44.jar:10.1.44]
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:903) ~[tomcat-embed-core-10.1.44.jar:10.1.44]
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1769) ~[tomcat-embed-core-10.1.44.jar:10.1.44]
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:52) ~[tomcat-embed-core-10.1.44.jar:10.1.44]
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1189) ~[tomcat-embed-core-10.1.44.jar:10.1.44]
	at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:658) ~[tomcat-embed-core-10.1.44.jar:10.1.44]
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:63) ~[tomcat-embed-core-10.1.44.jar:10.1.44]
	at java.base/java.lang.Thread.run(Thread.java:840) ~[na:na]
Caused by: java.lang.ArrayStoreException: org.eclipse.sirius.components.papaya.impl.ContainingLinkImpl
	at org.eclipse.emf.common.util.BasicEList.assign(BasicEList.java:117) ~[org.eclipse.emf.common-2.31.0.jar:na]
	at org.eclipse.emf.common.util.BasicEList.addUnique(BasicEList.java:416) ~[org.eclipse.emf.common-2.31.0.jar:na]
	at org.eclipse.emf.common.notify.impl.NotifyingListImpl.doAddUnique(NotifyingListImpl.java:325) ~[org.eclipse.emf.common-2.31.0.jar:na]
	at org.eclipse.emf.common.notify.impl.NotifyingListImpl.addUnique(NotifyingListImpl.java:282) ~[org.eclipse.emf.common-2.31.0.jar:na]
	at org.eclipse.emf.common.util.AbstractEList.add(AbstractEList.java:304) ~[org.eclipse.emf.common-2.31.0.jar:na]
	at org.eclipse.sirius.web.application.views.explorer.services.DefaultObjectDuplicator.lambda$updateIncomingReferences$4(DefaultObjectDuplicator.java:106) ~[classes/:na]
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183) ~[na:na]
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179) ~[na:na]
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625) ~[na:na]
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:762) ~[na:na]
	at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:276) ~[na:na]
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197) ~[na:na]
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179) ~[na:na]
	at java.base/java.util.AbstractList$RandomAccessSpliterator.forEachRemaining(AbstractList.java:720) ~[na:na]
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509) ~[na:na]
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499) ~[na:na]
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150) ~[na:na]
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173) ~[na:na]
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[na:na]
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596) ~[na:na]
	at org.eclipse.sirius.web.application.views.explorer.services.DefaultObjectDuplicator.updateIncomingReferences(DefaultObjectDuplicator.java:103) ~[classes/:na]
	at org.eclipse.sirius.web.application.views.explorer.services.DefaultObjectDuplicator.duplicateObject(DefaultObjectDuplicator.java:83) ~[classes/:na]
	at org.eclipse.sirius.web.application.views.explorer.services.ComposedObjectDuplicator.lambda$duplicateObject$2(ComposedObjectDuplicator.java:51) ~[classes/:na]
	at java.base/java.util.Optional.orElseGet(Optional.java:364) ~[na:na]
	at org.eclipse.sirius.web.application.views.explorer.services.ComposedObjectDuplicator.duplicateObject(ComposedObjectDuplicator.java:51) ~[classes/:na]
	at org.eclipse.sirius.web.application.views.explorer.services.DuplicateObjectEventHandler.duplicateObject(DuplicateObjectEventHandler.java:115) ~[classes/:na]
	at org.eclipse.sirius.web.application.views.explorer.services.DuplicateObjectEventHandler.handle(DuplicateObjectEventHandler.java:85) ~[classes/:na]
	at org.eclipse.sirius.components.collaborative.editingcontext.EditingContextEventProcessor.handleInput(EditingContextEventProcessor.java:324) ~[classes/:na]
	at org.eclipse.sirius.components.collaborative.editingcontext.EditingContextEventProcessor.doHandle(EditingContextEventProcessor.java:259) ~[classes/:na]
	at org.eclipse.sirius.components.collaborative.editingcontext.EditingContextEventProcessor.lambda$handle$3(EditingContextEventProcessor.java:226) ~[classes/:na]
	at org.eclipse.sirius.components.web.concurrent.DelegatingRequestContextRunnable.run(DelegatingRequestContextRunnable.java:40) ~[classes/:na]
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539) ~[na:na]
	at java.base/java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:264) ~[na:na]
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[na:na]
	... 1 common frames omitted
```

--- a/CHANGELOG.adoc
+++ b/CHANGELOG.adoc
@@ -27,6 +27,7 @@
 
 - https://github.com/eclipse-sirius/sirius-web/issues/5491[#5491] [test] Add missing module in tsconfig for the playwright package
 - https://github.com/eclipse-sirius/sirius-web/issues/5627[#5627] [sirius-web] Fix the duplication of an element with incoming non changeable references
+- https://github.com/eclipse-sirius/sirius-web/issues/5617[#5617] [sirius-web] Fix duplication of an element to a mono valued containment feature while updating incoming reference.
 
 
 === New Features

--- a/packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/DefaultObjectDuplicator.java
+++ b/packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/services/DefaultObjectDuplicator.java
@@ -65,13 +65,13 @@ public class DefaultObjectDuplicator implements IDefaultObjectDuplicator {
                     .filter(reference -> containmentFeature.equals(reference.getName()))
                     .findFirst();
 
-            Optional<Object> optionalDuplicatedObject = Optional.empty();
+            Optional<EObject> optionalDuplicatedObject = Optional.empty();
 
             if (optionalEReference.isPresent()) {
                 if (this.canFeatureBeSet(containerEObject, optionalEReference.get())) {
-                    optionalDuplicatedObject = this.duplicateObject(objectToDuplicate, settings)
-                            .flatMap(eObject -> this.addOrSetEObjet(optionalEditingDomain.get(), eObject, containerEObject, optionalEReference.get()));
-                    if (optionalDuplicatedObject.isPresent()) {
+                    optionalDuplicatedObject = this.duplicateObject(objectToDuplicate, settings);
+                    optionalDuplicatedObject.ifPresent(eObject -> this.addOrSetEObjet(optionalEditingDomain.get(), eObject, containerEObject, optionalEReference.get()));
+                    if (optionalDuplicatedObject.isPresent() && optionalDuplicatedObject.get().eContainer() != null) {
                         result = new Success(ChangeKind.SEMANTIC_CHANGE, Map.of(IObjectDuplicator.NEW_OBJECT, optionalDuplicatedObject.get()));
                     }
                 } else {
@@ -79,8 +79,8 @@ public class DefaultObjectDuplicator implements IDefaultObjectDuplicator {
                 }
             }
 
-            if (settings.updateIncomingReferences() && optionalDuplicatedObject.isPresent() && optionalDuplicatedObject.get() instanceof EObject duplicatedEObject) {
-                this.updateIncomingReferences(duplicatedEObject, objectToDuplicate);
+            if (settings.updateIncomingReferences() && optionalDuplicatedObject.isPresent()) {
+                this.updateIncomingReferences(optionalDuplicatedObject.get(), objectToDuplicate);
             }
         }
         return result;
@@ -111,7 +111,7 @@ public class DefaultObjectDuplicator implements IDefaultObjectDuplicator {
                 });
     }
 
-    private Optional<Object> addOrSetEObjet(EditingDomain editingDomain, EObject duplicatedObject, EObject container, EReference eReference) {
+    private void addOrSetEObjet(EditingDomain editingDomain, EObject duplicatedObject, EObject container, EReference eReference) {
         Command command;
         if (eReference.isMany()) {
             command = new AddCommand(editingDomain, container, eReference, List.of(duplicatedObject));
@@ -119,10 +119,7 @@ public class DefaultObjectDuplicator implements IDefaultObjectDuplicator {
             command = new SetCommand(editingDomain, container, eReference, duplicatedObject);
         }
         editingDomain.getCommandStack().execute(command);
-        if (command.getResult().size() == 1) {
-            return Optional.of(command.getResult().iterator().next());
-        }
-        return Optional.empty();
+
     }
 
     private Optional<EObject> duplicateObject(EObject eObjectToDuplicate, DuplicationSettings settings) {

